crates/pop-fork/src/cache.rs

```diff
@@ -0,0 +1,564 @@
+
+		Ok(Self { pool })
+	}
+
+	/// Get a cached storage value.
+	///
+	/// # Returns
+	/// * `Ok(Some(Some(value)))` - Cached with a value
+	/// * `Ok(Some(None))` - Cached as empty (storage key exists but has no value)
+	/// * `Ok(None)` - Not in cache (unknown)
+	pub async fn get(
+		&self,
+		block_hash: H256,
+		key: &[u8],
+	) -> Result<Option<Option<Vec<u8>>>, CacheError> {
+		// Retrieve the cached value and its empty flag for the given block and key.
+		// We need both `value` and `is_empty` to distinguish between:
+		// - Key not in cache (no row returned)
+		// - Key cached as empty (row exists, is_empty = true)
+		// - Key cached with value (row exists, is_empty = false)
+		let row =
+			sqlx::query("SELECT value, is_empty FROM storage WHERE block_hash = ? AND key = ?")
+				.bind(block_hash.as_bytes())
+				.bind(key)
+				.fetch_optional(&self.pool)
+				.await?;
+
+		Ok(row.map(|r| {
+			let is_empty: bool = r.get("is_empty");
+			if is_empty { None } else { Some(r.get("value")) }
+		}))
+	}
+
+	/// Cache a storage value.
+	///
+	/// # Arguments
+	/// * `block_hash` - The block hash this storage is from
+	/// * `key` - The storage key
+	/// * `value` - The storage value, or None if the key has no value (empty)
+	pub async fn set(
+		&self,
+		block_hash: H256,
+		key: &[u8],
+		value: Option<&[u8]>,
+	) -> Result<(), CacheError> {
+		// Insert or update the cached storage entry.
+		// Uses INSERT OR REPLACE (SQLite's UPSERT) to handle both new entries
+		// and updates to existing entries with the same (block_hash, key) primary key.
+		// The `is_empty` flag is set based on whether value is None, allowing us
+		// to cache the knowledge that a storage key has no value (vs not being cached).
+		sqlx::query(
+			"INSERT OR REPLACE INTO storage (block_hash, key, value, is_empty) VALUES (?, ?, ?, ?)",
+		)
+		.bind(block_hash.as_bytes())
+		.bind(key)
+		.bind(value)
+		.bind(value.is_none())
+		.execute(&self.pool)
+		.await?;
+
+		Ok(())
+	}
+
+	/// Get multiple cached storage values in a batch.
+	///
+	/// Returns results in the same order as the input keys.
+	pub async fn get_batch(
+		&self,
+		block_hash: H256,
+		keys: &[Vec<u8>],
+	) -> Result<Vec<Option<Option<Vec<u8>>>>, CacheError> {
+		if keys.is_empty() {
+			return Ok(vec![]);
+		}
+
+		// Build a SELECT query with dynamic IN clause for batch retrieval.
+		// This fetches all requested keys in a single round-trip to the database,
+		// which is more efficient than individual queries when fetching many keys.
+		// Example: SELECT ... WHERE block_hash = ? AND key IN (?, ?, ?)
+		let placeholders: Vec<_> = keys.iter().map(|_| "?").collect();
+		let query = format!(
+			"SELECT key, value, is_empty FROM storage WHERE block_hash = ? AND key IN ({})",
+			placeholders.join(", ")
+		);
+
+		let mut query_builder = sqlx::query(&query).bind(block_hash.as_bytes());
+
+		for key in keys {
+			query_builder = query_builder.bind(key.as_slice());
```

Comment:
Just to apply if the previous suggestion is also applied

```suggestion
			query_builder = query_builder.bind(key);
```